### PR TITLE
SONAR-6682 Move computation of comment density measures to Compute Engine

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/formula/CoreFormulaRepositoryImpl.java [hunk 1]
```diff
@@ -41,8 +42,8 @@
 public class CoreFormulaRepositoryImpl implements CoreFormulaRepository {
 
   private static final List<Formula> FORMULAS = ImmutableList.<Formula>of(
-    // TODO When all decorators will be moved to CE, uncomment commented lines to activate all formulas and remove formulas declaration in
-    // {@link org.sonar.api.measures.CoreMetrics}
+    // TODO When all decorators will be moved to CE, uncomment commented lines to activate all formulas and remove formulas declaration in {@link org.sonar.api.measures.CoreMetrics}
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/formula/CoreFormulaRepositoryImpl.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import static org.sonar.api.measures.CoreMetrics.COMMENT_LINES_KEY;
+import static org.sonar.api.measures.CoreMetrics.NCLOC_KEY;
+
+/**
+ * Computes comments measures on files and then aggregates them on higher components.
+ */
+public class CommentMeasuresStep implements ComputationStep {
+
+  private static final ImmutableList<Formula> FORMULAS = ImmutableList.<Formula>of(
+    // TODO add this formula when {@link CoreMetrics.DUPLICATED_LINES_DENSITY} will be computed in CE
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/formula/CoreFormulaRepositoryImpl.java [hunk 3]
```diff
@@ -51,12 +52,14 @@
     new SumFormula(CLASSES_KEY),
     new SumFormula(FUNCTIONS_KEY),
     new SumFormula(STATEMENTS_KEY),
-    // new SumFormula(PUBLIC_API_KEY),
-    // new SumFormula(COMMENT_LINES_KEY),
-    // new SumFormula(PUBLIC_UNDOCUMENTED_API_KEY),
     new SumFormula(COMPLEXITY_KEY),
     new SumFormula(COMPLEXITY_IN_CLASSES_KEY),
+    // TODO this formula seems to be useless as this measure seems only required on files
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/formula/CoreFormulaRepositoryImpl.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by sns-seb** [comment]
You can refactor your Formulas in both step to a single Formula in each step which will generate more than one measure
therefor, you need the multimeasure support Formula available in #415 
in the meantime, the PR is yours

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by julienlancelot** [comment]
Feedback apply @sns-seb !

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      return new BiSumCounter(PUBLIC_API_KEY, PUBLIC_UNDOCUMENTED_API_KEY);
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(BiSumCounter counter, CreateMeasureContext context) {
+      if (counter.getValue1().isPresent() && counter.getValue2().isPresent()) {
+        int publicApis = counter.getValue1().get();
+        int publicUndocumentedApis = counter.getValue2().get();
+        if (publicApis > 0d) {
+          double documentedAPI = publicApis - publicUndocumentedApis;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java]: MAJOR Cast one of the operands of this substraction operation to a "double". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    public BiSumCounter createNewCounter() {
+      return new BiSumCounter(NCLOC_KEY, COMMENT_LINES_KEY);
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(BiSumCounter counter, CreateMeasureContext context) {
+      if (counter.getValue1().isPresent() && counter.getValue2().isPresent()) {
+        int nclocs = counter.getValue1().get();
+        int comments = counter.getValue2().get();
+        double divisor = nclocs + comments;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java]: MAJOR Cast one of the operands of this addition operation to a "double". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      return new BiSumCounter(PUBLIC_API_KEY, PUBLIC_UNDOCUMENTED_API_KEY);
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(BiSumCounter counter, CreateMeasureContext context) {
+      if (counter.getValue1().isPresent() && counter.getValue2().isPresent()) {
+        int publicApis = counter.getValue1().get();
+        int publicUndocumentedApis = counter.getValue2().get();
+        if (publicApis > 0) {
+          double documentedAPI = publicApis - publicUndocumentedApis;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java]: MAJOR Cast one of the operands of this substraction operation to a "double". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      return new BiSumCounter(PUBLIC_API_KEY, PUBLIC_UNDOCUMENTED_API_KEY);
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(BiSumCounter counter, CreateMeasureContext context) {
+      if (counter.getValue1().isPresent() && counter.getValue2().isPresent()) {
+        double publicApis = counter.getValue1().get();
+        double publicUndocumentedApis = counter.getValue2().get();
+        if (publicApis > 0d) {
+          double documentedAPI = publicApis - publicUndocumentedApis;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/DocumentationMeasuresStep.java]: MAJOR Cast one of the operands of this substraction operation to a "double". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    public BiSumCounter createNewCounter() {
+      return new BiSumCounter(NCLOC_KEY, COMMENT_LINES_KEY);
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(BiSumCounter counter, CreateMeasureContext context) {
+      if (counter.getValue1().isPresent() && counter.getValue2().isPresent()) {
+        double nclocs = counter.getValue1().get();
+        double comments = counter.getValue2().get();
+        double divisor = nclocs + comments;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/CommentMeasuresStep.java]: MAJOR Cast one of the operands of this addition operation to a "double". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2184)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Pull request merged**